sql/export: add header row option to export csv

log-head opened the pull request:
EXPORT currently does not include column headers, which can lead to issues when columns do not have the same ordinal number (i.e., order) in the database schema that the export is being loaded to. By including column headers, users will be able to disambiguate the columns in the export file.
    
Fixes: #143884
Epic: CRDB-49136
    
Release note (sql change): WITH header_row flag is added to EXPORT. Returns error for non-csv type. Another row is prepended to the csv file with the column names.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/149686)
<!-- Reviewable:end -->

asg0451 commented on pkg/sql/execinfrapb/processors_bulk_io.proto:
```diff
@@ -401,6 +401,9 @@ message ExportSpec {
   repeated string col_names = 7 ;
+
+  // header_row specifies if a csv file should include header rows
+  optional bool header_row = 8 ;
```
nit: can you use gogoproto.nullable as above to make this not have to be a ptr?

asg0451 commented on pkg/sql/export.go:
```diff
@@ -196,6 +200,10 @@ func buildExportPlanningInfo(
 	exprEval := planner.ExprEvaluator("EXPORT")
 	treeOptions := make(tree.KVOptions, len(options))
 	for i, o := range options {
+		if o.Value.(*tree.CastExpr).Expr == tree.DNull {
```
is this type assertion always valid? if not, do an `, ok ` check. i dont know enough about this code to say one way or the other.

asg0451 commented on pkg/sql/export.go:
```diff
@@ -196,6 +200,10 @@ func buildExportPlanningInfo(
 	treeOptions := make(tree.KVOptions, len(options))
 	for i, o := range options {
+		if o.Value.(*tree.CastExpr).Expr == tree.DNull {
+			o.Value = nil
```
i'd prefer not to mutate `options` here. can you do this in another way that doesnt? maybe setting nil on `treeOptions[i]`

asg0451 commented on pkg/sql/export/exportcsv.go:
```diff
@@ -178,6 +178,11 @@ func (sp *csvWriter) Run(ctx context.Context, output execinfra.RowReceiver) {
 
 		writer := newCSVExporter(sp.spec)
+		if sp.spec.HeaderRow != nil && *sp.spec.HeaderRow {
+			if err := writer.Write(sp.spec.ColNames); err != nil {
```
can you test this with double-quoted-named columns? eg `"a"`, `"A"`, `"a""B"`, etc

asg0451 commented on pkg/sql/export/exportcsv_test.go:
```diff
@@ -791,3 +791,46 @@ func TestProcessorEncountersUncertaintyError(t *testing.T) {
+		}
+	})
+
+	t.Run("header row with nullas", func(t *testing.T) {
```
i dont think this subtest actually tests nullas. that's probably fine, but you should fix that or remove the subtest

asg0451 commented on pkg/sql/export/exportcsv_test.go:
```diff
@@ -791,3 +791,46 @@ func TestProcessorEncountersUncertaintyError(t *testing.T) {
+		sqlDB.Exec(t, `EXPORT INTO CSV 'nodelocal://1/header_nullas' WITH header_row, nullas = '' FROM SELECT * FROM foo`)
+		content := readFileByGlob(t, filepath.Join(dir, "header_nullas", exportFilePattern))
+
+		if expected, got := "i,x,y,z\n1,12,3,14\n", string(content); expected != got {
+			t.Fatalf("expected %q, got %q", expected, got)
+		}
+	})
```
nit; here and elsewhere: 
```suggestion
		require.Equal(t, "i,x,....", string(content))
```

asg0451 commented on pkg/sql/export/exportcsv_test.go:
```diff
@@ -791,3 +791,46 @@ func TestProcessorEncountersUncertaintyError(t *testing.T) {
+		sqlDB.Exec(t, `EXPORT INTO CSV 'nodelocal://1/header_quotes' WITH header_row, nullas = '' FROM SELECT * FROM bar`)
+		content := readFileByGlob(t, filepath.Join(dir, "header_quotes", exportFilePattern))
+
+		if expected, got := "a,b,c ,D,\"e\"\"f\"\n1,12,3,14,56\n", string(content); expected != got {
```
nit: can you use backticks to make this more readable?

asg0451 reviewed (approved):
lgtm! just a few testing nits to fix

yuzefovich reviewed (commented):
> Reviewed 2 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149686)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao, @log-head, and @rharding6373)

___
*[`pkg/sql/execinfrapb/processors_bulk_io.proto` line 386 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/149686#-OUqLcUGBzMZp2wF2JMD:-OUqLcUGBzMZp2wF2JME:b-bi1pi) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4e2fc1070c8a0d7dc54f40b45464e86a2c9169e/pkg/sql/execinfrapb/processors_bulk_io.proto#L386)):*
> ```proto
> // writes them to Parquet or CSV files at uri. It outputs a row per file written with
> // the file name, row count and byte size.
> message ExportSpec {
> ```

nit: it'd be good to move `ExportSpec` into a proto file owned by CDC to avoid pinging DR on changes to it. Probably a separate `_export.proto` file would be best.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
I'd suggest making new `pkg/sql/execinfrapb/processors_export.proto` file which would contain the ExportSpec and then mark it as owned by CDC in `.CODEOWNERS`. (Take a look at `processors_inspect.proto` for a recent example.)

yuzefovich reviewed (commented):
> Reviewed 4 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149686)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao, @log-head, and @rharding6373)

___
*[`.github/CODEOWNERS` line 65 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149686#-OUvTNRX9_OWWkr3jzjI:-OUvTNRX9_OWWkr3jzjJ:b3jtzma) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4530d7a011fcc04c53258496fed4d5014720ed4/.github/CODEOWNERS#L65)):*
> ```CODEOWNERS
> /pkg/sql/execinfrapb/processors_changefeeds.proto @cockroachdb/cdc-prs
> /pkg/sql/execinfrapb/processors_ttl.proto         @cockroachdb/sql-foundations
> /pkg/sql/export/processors_export.proto         @cockroachdb/cdc-prs
> ```

We need to keep the correct file path: `/pkg/sql/execinfrapb/processors_export.proto`.


<!-- Sent from Reviewable.io -->

andyyang890 commented on pkg/sql/execinfrapb/processors_export.proto:
```diff
@@ -0,0 +1,45 @@
+// docs/RFCS/distributed_sql.md.
+// All the concepts here are "physical plan" concepts.
+
+syntax = "proto2";
```
nit: I'd prefer to see this code move (and CODEOWNERS file change) in a separate commit

log-head committed:
sql/execinfrapb: extracting export proto

Previously, the export proto was shared in processors_bulk_io.proto
and is assigned to disaster recovery team. Moving the proto definitions
for export to assign to cdc.

Epic: none
Fixes: none
Release note: None

log-head committed:
sql/export: add header row option to export csv

EXPORT currently does not include column headers, which can lead to
issues when columns do not have the same ordinal number (i.e., order) in
the database schema that the export is being loaded to. By including
column headers, users will be able to disambiguate the columns in the
export file.

Fixes: #143884
Epic: CRDB-49136

Release note (sql change): WITH header_row flag is added to EXPORT.
Returns error for non-csv type. Another row is prepended to the csv file
with the column names.

log-head commented:
bors r-

andyyang890 reviewed (approved):
Nice work!
